src/OrleansRuntime/Storage/MemoryStorageGrain.cs

```diff
@@ -119,23 +119,23 @@ private static string NewEtag()
 
             private void ValidateEtag(string currentETag, string receivedEtag, string grainStoreKey, string operation)
             {
+                // if we have no current etag, we will except the users.
+                // This is a mitigation for when the memory storage grain is lost do to silo crash.
+                if (currentETag == null)
+                    return;
```

Comment:
This is a little too optimistic. Deactivation (crash) of the grain is expected but uncommon, so we should just honor the new state if this is a new activation (I agree with the usefulness of being less strict after failure).
Nevertheless, the current approach wouldn't catch the following (non-failure) scenario, that could be common:
1. grain1 reads state with etag X
2. grain2 reads state with etag X
3. grain 2 clears state using etag X (etag gets wiped)
4. grain1 updates state using etag X (because etag is null, the current implementation will accept the change, but it shouldn't).

In my opinion we should encode the MemoryStorageGrain activation ID in the eTag (which should be opaque to the callers anyway). Then, if the situation occurs that an eTag is provided, but the store has a null eTag for that state, then check whether the original eTag was generated by the same activation of the MemoryStoreGrain, or this happens to be a new activation (failure tolerance that this PR intends to tackle). Only be less strict if a different activation generated the previous eTag